File: Low Level Design (LLD)/Problems/06. Splitwise/Java/model/PercentageExpense.java
package model;

import model.enums.ExpenseType;

import java.util.*;

public class PercentageExpense extends Expense {
    private List<Double> percentages;

    public PercentageExpense(String expenseId, String description, double amount, User paidBy, List<User> participants, List<Double> percentages) {
        super(expenseId, description, amount, paidBy, participants, ExpenseType.PERCENTAGE);
        this.percentages = percentages;
    }

    @Override
    public void calculateSplit() {
        for (int i = 0; i < participants.size(); i++) {
            double share = (percentages.get(i) / 100.0) * amount;
            userShare.put(participants.get(i), share);
        }
    }
}

File: Low Level Design (LLD)/Problems/06. Splitwise/Cpp/model/src/PercentageExpense.cpp
#include "../include/PercentageExpense.h"

PercentageExpense::PercentageExpense(const std::string& expenseId,
                                     const std::string& description,
                                     double amount,
                                     std::shared_ptr<User> paidBy,
                                     const std::vector<std::shared_ptr<User>>& participants,
                                     const std::vector<double>& percentages)
    : Expense(expenseId, description, amount, paidBy, participants, ExpenseType::PERCENTAGE),
      percentages(percentages) {}

void PercentageExpense::calculateSplit() {
    auto& userShareMap = getUserShareMutable();

    for (size_t i = 0; i < participants.size(); ++i) {
        double share = (percentages[i] / 100.0) * amount;
        userShareMap[participants[i]] = share;
    }
}
